numpy/core/tests/test_multiarray.py

```diff
@@ -6021,7 +6022,47 @@ def test_dot_equivalent(self, args):
 
         r3 = np.matmul(args[0].copy(), args[1].copy())
         assert_equal(r1, r3)
-        
+    
+    def test_matmul_object(self):
+        import fractions
+
+        f = np.vectorize(fractions.Fraction)
+        def random_ints():
+            return np.random.randint(1, 1000, size=(10, 3, 3))
+        M1 = f(random_ints(), random_ints()) 
+        M2 = f(random_ints(), random_ints())
+
+        M3 = self.matmul(M1, M2)
+
+        [N1, N2, N3] = [a.astype(float) for a in [M1, M2, M3]]
+
+        assert_allclose(N3, self.matmul(N1, N2))
+
+    def test_matmul_object_type_scalar(self):
+        from fractions import Fraction as F
+        v = np.array([F(2,3), F(5,7)])
+        res = self.matmul(v, v)
+        assert_(type(res) is F)
+
+    def test_matmul_empty(self):
+        a = np.empty((3, 0), dtype=object)
+        b = np.empty((0, 3), dtype=object)
+        c = np.zeros((3, 3))
+        assert_array_equal(np.matmul(a, b), c)
+
+    def test_matmul_exception_multiply(self):
+        with assert_raises(TypeError):
+            a = np.full((3,3), None)  
+            b = np.matmul(a, a)
+
+    def test_matmul_exception_add(self):
+        class multiply_not_add():
+            def __mul__(self, other):
+                return self
+        with assert_raises(TypeError):
+            a = np.full((3,3), multiply_not_add())
```

Comment:
As above.

Also worth adding: a comment on this test saying something like ``# test that matmul fails if `__add__` is missing``